File: binary_edition/backend/binary_parser.py
"""
Binary Telemetry Parser (64-Byte Struct)
For GPS Tambang Edge - Binary Edition
"""

import struct
from datetime import datetime, timezone
from typing import Optional, Dict, Any

TELEMETRY_STRUCT_FMT_V2 = "<2sB8sQIIiiHHhHBBB6sbIBhhhH"
TELEMETRY_STRUCT_FMT_V1 = "<2sB8sIIiiHHhHI4Bh3h6sbBBHH"
TELEMETRY_PACKET_SIZE = 66

def calculate_crc16(data: bytes) -> int:
    """CRC16-CCITT (Poly: 0x1021, Init: 0xFFFF) matching ESP32 implementation."""
    crc = 0xFFFF
    for byte in data:
        crc ^= (byte << 8)
        for _ in range(8):
            if crc & 0x8000:
                crc = ((crc << 1) ^ 0x1021) & 0xFFFF
            else:
                crc = (crc << 1) & 0xFFFF
    return crc

def parse_telemetry_packet(raw_bytes: bytes) -> Optional[Dict[str, Any]]:
    """
    Unpacks 66-byte raw telemetry binary packet into a clean Python dictionary.
    Returns None if packet is invalid or CRC checksum fails.
    """
    if len(raw_bytes) != TELEMETRY_PACKET_SIZE:
        return None

    if raw_bytes[:2] != b'\xaa\x55':
        return None

    # Validate CRC16 over first 64 bytes
    received_crc = struct.unpack("<H", raw_bytes[64:66])[0]
    calculated_crc = calculate_crc16(raw_bytes[:64])
    if calculated_crc != received_crc:
        return None

    version = raw_bytes[2]
    imei_str = ""
    ibutton_hex = ""
    ibutton_id = 0
    ibutton_flags = 0
    output_status = 0
    hdop = 0.0
    temp = 0.0
    gs_x, gs_y, gs_z = 0, 0, 0
    event_code = 51

    if version == 2:
        (
            magic,
            version,
            src_raw,
            imei_num,
            seq,
            timestamp_sec,
            lat_x1e7,
            lon_x1e7,
            speed_x10,
            heading,
            altitude,
            bat_mv,
            ignition,
            input_status,
            flags,
            beacon_mac_raw,
            beacon_rssi,
            ibutton_id,
            ibutton_flags,
            gs_x,
            gs_y,
            gs_z,
            received_crc
        ) = struct.unpack(TELEMETRY_STRUCT_FMT_V2, raw_bytes)
        imei_str = str(imei_num) if imei_num > 0 else ""
        if ibutton_id > 0:
            ibutton_hex = f"{ibutton_id:08X}"
    else:
        (
            magic,
            version,
            src_raw,
            seq,
            timestamp_sec,
            lat_x1e7,
            lon_x1e7,
            speed_x10,
            heading,
            altitude,
            bat_mv,
            odo_m,
            ignition,
            input_status,
            output_status,
            hdop_x10,
            temp_x10,
            gs_x,
            gs_y,
            gs_z,
            beacon_mac_raw,
            beacon_rssi,
            event_code,
            flags,
            reserved,
            received_crc
        ) = struct.unpack(TELEMETRY_STRUCT_FMT_V1, raw_bytes)
        hdop = round(hdop_x10 / 10.0, 1)
        temp = round(temp_x10 / 10.0, 1)

    # Decode string source ID
    src = src_raw.decode('ascii', errors='ignore').rstrip('\x00').strip()
    
    # Format Beacon MAC Address
    if any(b != 0 for b in beacon_mac_raw):
        beacon_mac = ":".join(f"{b:02X}" for b in beacon_mac_raw)
    else:
        beacon_mac = ""

    # 🛡️ Data Integrity Guard: Tolak jika timestamp invalid, lat/lon 0.0, atau IMEI kosong
    if timestamp_sec < 1577836800 or (lat_x1e7 == 0 and lon_x1e7 == 0) or not imei_str or imei_str == "0":
        return None

    # Convert Timestamp to ISO8601 string
    try:
        dt = datetime.fromtimestamp(timestamp_sec, tz=timezone.utc)
        ts_iso = dt.strftime("%Y-%m-%d %H:%M:%S")
    except Exception:
        ts_iso = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S")

    unique_id = f"{src}-{timestamp_sec}-{seq}"

    return {
        "id": unique_id,
        "src": src,
        "imei": imei_str,
        "ibutton": ibutton_hex,
        "ibutton_auth": bool(ibutton_flags & 0x02),
        "ibutton_login": bool(ibutton_flags & 0x01),
        "seq": seq,
        "timestamp_sec": timestamp_sec,
        "ts": ts_iso,
        "lat": round(lat_x1e7 / 1e7, 7),
        "lon": round(lon_x1e7 / 1e7, 7),
        "spd": round(speed_x10 / 10.0, 1),
        "hdg": heading,
        "alt": altitude,
        "bat": round(bat_mv / 1000.0, 2),
        "bat_mv": bat_mv,
        "ign": ignition,
        "pto": 1 if (input_status & 0x01) else 0,
        "in": f"{input_status:02X}",
        "out": f"{output_status:02X}",
        "hdop": hdop,
        "temp": temp,
        "gs": {
            "x": gs_x,
            "y": gs_y,
            "z": gs_z
        },
        "be": [
            {
                "mac": beacon_mac,
                "rssi": beacon_rssi
            }
        ] if beacon_mac else [],
        "event_code": event_code,
        "flags": flags
    }

def parse_telemetry_batch(raw_bytes: bytes) -> list:
    """
    Unpacks a raw payload containing 1 or more 64-byte binary packets.
    Returns a list of valid parsed telemetry dictionaries.
    """
    if not raw_bytes:
        return []
    
    results = []
    num_packets = len(raw_bytes) // TELEMETRY_PACKET_SIZE
    for i in range(num_packets):
        chunk = raw_bytes[i * TELEMETRY_PACKET_SIZE : (i + 1) * TELEMETRY_PACKET_SIZE]
        parsed = parse_telemetry_packet(chunk)
        if parsed:
            results.append(parsed)
    return results

if __name__ == "__main__":
    # Test pack and unpack
    src_b = b"EXCA01\x00\x00"
    mac_b = bytes([0xC3, 0x00, 0x00, 0x38, 0xB4, 0x52])
    test_data = struct.pack(
        "<2sB8sIIiiHHhHI4Bh3h6sbBBHH",
        b'\xaa\x55', 1, src_b, 100, 1775738126,
        int(-0.7388810 * 1e7), int(117.1301520 * 1e7),
        250, 67, 45, 25400, 3815, 1, 0, 0, 8, 425,
        -734, -263, 583, mac_b, -67, 51, 1, 0, 0
    )
    # Calculate and put CRC
    crc = calculate_crc16(test_data[:62])
    final_packet = test_data[:62] + struct.pack("<H", crc)
    
    # Test batch unpack with 2 packets
    batch_data = final_packet * 2
    parsed_batch = parse_telemetry_batch(batch_data)
    print(f"Test Batch Unpack: Parsed {len(parsed_batch)} packets.")
    import pprint
    pprint.pprint(parsed_batch[0])

File: dashboard realtime binary/binary_parser.js
/**
 * Binary Parser for GPS Tambang Telemetry (66-Byte Struct Version 2)
 * Matches firmware TelemetryPacketBinary exactly.
 */

const TELEMETRY_PACKET_SIZE = 66;

// CRC16-CCITT (Poly: 0x1021, Init: 0xFFFF)
function calculateCRC16(buffer, length) {
  let crc = 0xFFFF;
  for (let i = 0; i < length; i++) {
    crc ^= (buffer[i] << 8);
    for (let j = 0; j < 8; j++) {
      if ((crc & 0x8000) !== 0) {
        crc = ((crc << 1) ^ 0x1021) & 0xFFFF;
      } else {
        crc = (crc << 1) & 0xFFFF;
      }
    }
  }
  return crc;
}

function parseBinaryPacket(buf) {
  if (buf.length !== TELEMETRY_PACKET_SIZE) return null;
  if (buf[0] !== 0xAA || buf[1] !== 0x55) return null;

  // CRC Check
  const receivedCrc = buf.readUInt16LE(64);
  const calculatedCrc = calculateCRC16(buf, 64);
  if (receivedCrc !== calculatedCrc) {
    return null;
  }

  const version = buf.readUInt8(2);

  // Device ID (bytes 3..10, 8 bytes null-terminated ASCII)
  let src = buf.subarray(3, 11).toString('ascii').replace(/\0/g, '').trim();
  if (!src) src = "UNKNOWN";

  // IMEI (bytes 11..18, uint64 Little Endian)
  let imei = "";
  try {
    const imeiBig = buf.readBigUInt64LE(11);
    imei = imeiBig > 0n ? imeiBig.toString() : "";
  } catch (e) {
    imei = "";
  }

  const seq = buf.readUInt32LE(19);
  const timestamp = buf.readUInt32LE(23);
  const lat_x1e7 = buf.readInt32LE(27);
  const lon_x1e7 = buf.readInt32LE(31);
  const speed_x10 = buf.readUInt16LE(35);
  const heading = buf.readUInt16LE(37);
  const altitude = buf.readInt16LE(39);
  const bat_mv = buf.readUInt16LE(41);
  const ignition = buf.readUInt8(43);
  const input_status = buf.readUInt8(44);
  const flags = buf.readUInt8(45);

  // Beacon MAC (bytes 46..51)
  const beaconBytes = buf.subarray(46, 52);
  let beaconMac = "";
  let hasBeacon = false;
  for (let i = 0; i < 6; i++) {
    if (beaconBytes[i] !== 0) {
      hasBeacon = true;
      break;
    }
  }
  if (hasBeacon) {
    beaconMac = Array.from(beaconBytes).map(b => b.toString(16).padStart(2, '0').toUpperCase()).join(':');
  }

  const beacon_rssi = buf.readInt8(52);
  const ibutton_id = buf.readUInt32LE(53);
  const ibutton_flags = buf.readUInt8(57);
  const gs_x = buf.readInt16LE(58);
  const gs_y = buf.readInt16LE(60);
  const gs_z = buf.readInt16LE(62);

  // Reject zero coordinate
  if (lat_x1e7 === 0 && lon_x1e7 === 0) return null;

  const lat = Math.round((lat_x1e7 / 1e7) * 1e7) / 1e7;
  const lon = Math.round((lon_x1e7 / 1e7) * 1e7) / 1e7;
  const spd = Math.round((speed_x10 / 10.0) * 10) / 10;
  const bat = Math.round((bat_mv / 1000.0) * 100) / 100;

  // Format ISO timestamp
  let dateObj = new Date(timestamp * 1000);
  if (isNaN(dateObj.getTime()) || timestamp < 1577836800) {
    dateObj = new Date();
  }
  const ts = dateObj.toISOString().replace('T', ' ').substring(0, 19);

  return {
    id: `${src}-${timestamp}-${seq}`,
    src: src,
    type: src.toUpperCase().startsWith('EXCA') ? 'EXCA' : 'DT',
    imei: imei,
    seq: seq,
    timestamp: timestamp,
    ts: ts,
    lat: lat,
    lon: lon,
    spd: spd,
    hdg: heading,
    alt: altitude,
    bat: bat,
    bat_mv: bat_mv,
    ign: ignition,
    pto: (input_status & 0x01) ? 1 : 0,
    gps_fix: (flags & 0x01) ? 1 : 0,
    relay: (flags & 0x02) ? 1 : 0,
    beacon_mac: beaconMac,
    beacon_rssi: beacon_rssi,
    ibutton: ibutton_id > 0 ? ibutton_id.toString(16).toUpperCase().padStart(8, '0') : "",
    ibutton_auth: (ibutton_flags & 0x02) ? true : false,
    gs: { x: gs_x, y: gs_y, z: gs_z },
    updated_at: Date.now()
  };
}

function parseBinaryPayload(payload) {
  if (!payload || payload.length === 0) return [];
  const records = [];
  const count = Math.floor(payload.length / TELEMETRY_PACKET_SIZE);

  for (let i = 0; i < count; i++) {
    const chunk = payload.subarray(i * TELEMETRY_PACKET_SIZE, (i + 1) * TELEMETRY_PACKET_SIZE);
    const rec = parseBinaryPacket(chunk);
    if (rec) {
      records.push(rec);
    }
  }
  return records;
}

module.exports = {
  TELEMETRY_PACKET_SIZE,
  parseBinaryPacket,
  parseBinaryPayload
};
